### PR TITLE
Explore removing use of `int` within header files to align with guidelines.

### PR DESCRIPTION
**Pull request by ahsonkhan**
Part of https://github.com/Azure/azure-sdk-for-cpp/issues/2226

Trying to focus mainly on headers, depending on the conclusion, we'd want to make cpp file changes as well, but that can be done later.

Assuming this is the guideline we want:
https://azure.github.io/azure-sdk/cpp_implementation.html#integer-sizes

image

Some places that we still use `int` is for things like unix file desriptor or where APIs like BIO_read from OpenSSL returns an int. Are those exceptions OK?
https://github.com/Azure/azure-sdk-for-cpp/blob/0b64ab3845d6850bc1666337f21640899e732ef5/sdk/core/azure-core/inc/azure/core/io/body_stream.hpp#L166
https://github.com/Azure/azure-sdk-for-cpp/blob/308e5363adaa00ace2f06ff0f1a31747a0a28062/sdk/core/azure-core/src/base64.cpp#L87

### CODE REVIEW

**Comment by ahsonkhan** [comment]
cc @CaseyCarter and @BillyONeal

File: sdk/core/azure-core/src/private/package_version.hpp [hunk 1]
```diff
@@ -24,13 +26,13 @@ namespace Azure { namespace Core { namespace _detail {
   class PackageVersion final {
   public:
     /// Major numeric identifier.
-    static constexpr int Major = AZURE_CORE_VERSION_MAJOR;
+    static constexpr int32_t Major = AZURE_CORE_VERSION_MAJOR;
```
Comment (ahsonkhan) [inline sdk/core/azure-core/src/private/package_version.hpp:29]: Didn't modify storage, but if we do this, we'd want to make the same change there.

**Review by Jinming-Hu — APPROVED** [review]
Really don't like this. I believe we should use `int` for those cases we know the value isn't too big. This is suggested by Google C++ Style Guide.

The rule in our C++ guidelines doesn't make any sense to me. But I'll accept it as it's guideline.

**Comment by ahsonkhan** [comment]
> This is suggested by Google C++ Style Guide.

For reference, I believe this is what @Jinming-Hu is referring to:
https://google.github.io/styleguide/cppguide.html#Integer_Types

image

**Comment by CaseyCarter** [comment]
"Don't use `int` under any circumstances" is a silly guideline that I would ignore. Presumably this guideline was originally formulated as a prohibition against using `int` to index into or represent the size of memory buffers and containers - which _is_ sensible given that such sizes may be larger than an `int` can represent - and then "simplified" into its current state.

You can't even declare `main` without using `int`.

**Comment by ahsonkhan** [comment]
> You can't even declare `main` without using `int`.

Haha, exactly :)

**Comment by BillyONeal** [comment]
> "Don't use `int` under any circumstances" is a silly guideline that I would ignore. Presumably this guideline was originally formulated as a prohibition against using `int` to index into or represent the size of memory buffers and containers - which _is_ sensible given that such sizes may be larger than an `int` can represent - and then "simplified" into its current state.
> 
> You can't even declare `main` without using `int`.

The intent was that the couple of places it would be necessary would require review and we would put that in the guidelines. In particular, int shouldn't be in any public API surface, for indexing / loops, or in any data structure defined by the SDK, when used as an int. (e.g. POSIX fds are `int` but they aren't used that way by the application).

Notably, anything that might go over the wire needs to be int64_t in case the service(s) are changed to allow that, as happened when Storage adopted Jumbo Blobs. (That requirement came from Jeffrey Richter)

I think the current list of known exceptions is:

* POSIX fds, or other situations where the int is an opaque handle
* Interfaces that are forced to be defined in terms of int, like main
* SFIANE/enable_if as the =0 target in a template

and the list of "absolutely problematic" ones are:

* Any public API surface (should probably be int64_t)
* Memory buffer sizes (should be size_t or ptrdiff_t)
* File sizes (should be int64_t)

> Really don't like this. I believe we should use `int` for those cases we know the value isn't too big. This is suggested by Google C++ Style Guide.
> 
> The rule in our C++ guidelines doesn't make any sense to me. But I'll accept it as it's guideline.

Our experience is that people think the values aren't too big, until the values can be too big, and then it's a nightmare to fix later. For example, this is exactly what happened in cpprestsdk that caused it to fall over with inputs bigger than 2GiB.

> > This is suggested by Google C++ Style Guide.
> 
> For reference, I believe this is what @Jinming-Hu is referring to:
> https://google.github.io/styleguide/cppguide.html#Integer_Types

I disagree with Google's style guide here -- note that their style guide effectively assumes an LP64 platform (Linux). Their justification is "uniformity of declaration," that is, to be consistent with the large base of existing code already present at Google, and I don't believe it's particularly relevant here.

**Comment by ahsonkhan** [comment]
> In particular, int shouldn't be in any public API surface, for indexing / loops, or in any data structure defined by the SDK, when used as an int. (e.g. POSIX fds are `int` but they aren't used that way by the application).

There are a couple of places we need to go fix. I will follow up and file issues.
https://github.com/Azure/azure-sdk-for-cpp/blob/e4a18b5eb3b556fae7bde96c43444f27f2fcbdb8/sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/share_options.hpp#L638-L642

**Review by antkmsft — APPROVED** [review]

File: sdk/core/azure-core/inc/azure/core/uuid.hpp [hunk 2]
```diff
@@ -28,7 +28,7 @@ namespace Azure { namespace Core {
   class Uuid final {
 
   private:
-    static constexpr int UuidSize = 16;
+    static constexpr int32_t UuidSize = 16;
```
Comment (CaseyCarter) [inline sdk/core/azure-core/inc/azure/core/uuid.hpp:31]: Why not `size_t` here as well? Isn't this used as the size of memory buffers?

**Review by vhvb1989 — APPROVED** [review]

**Pull request merged**